File: frontend/src/api/content.ts
import apiClient from './index'

const api = apiClient

export interface Content {
  id: number
  title: string
  content_markdown: string
  content_html: string
  source_type: string
  source_file: string | null
  author: string
  tags: string[]
  category: string
  cover_image: string | null
  status: string
  work_status: string
  owner_id: number | null
  community_id: number
  created_by_user_id: number | null
  assignee_ids: number[]
  community_ids: number[]
  scheduled_publish_at: string | null
  created_at: string
  updated_at: string
}

export interface ContentListItem {
  id: number
  title: string
  source_type: string
  author: string
  tags: string[]
  category: string
  status: string
  work_status: string
  community_id: number | null
  owner_id: number | null
  scheduled_publish_at: string | null
  created_at: string
  updated_at: string
  assignee_names: string[]
}

export interface ContentCalendarItem {
  id: number
  title: string
  status: string
  source_type: string
  author: string
  category: string
  scheduled_publish_at: string | null
  created_at: string
}

export interface PaginatedContents {
  items: ContentListItem[]
  total: number
  page: number
  page_size: number
}

export async function fetchContents(params: {
  page?: number
  page_size?: number
  status?: string
  source_type?: string
  keyword?: string
  community_id?: number
  unscheduled?: boolean
}): Promise<PaginatedContents> {
  const { data } = await api.get('/contents', { params })
  return data
}

export async function fetchContent(id: number): Promise<Content> {
  const { data } = await api.get(`/contents/${id}`)
  return data
}

export async function createContent(payload: Partial<Content>): Promise<Content> {
  const { data } = await api.post('/contents', payload)
  return data
}

export async function updateContent(id: number, payload: Partial<Content>): Promise<Content> {
  const { data } = await api.put(`/contents/${id}`, payload)
  return data
}

export async function deleteContent(id: number): Promise<void> {
  await api.delete(`/contents/${id}`)
}

export async function updateContentStatus(id: number, status: string): Promise<Content> {
  const { data } = await api.patch(`/contents/${id}/status`, { status })
  return data
}

export async function uploadFile(file: File): Promise<Content> {
  const formData = new FormData()
  formData.append('file', file)
  const { data } = await api.post('/contents/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export async function uploadCoverImage(contentId: number, file: File): Promise<Content> {
  const formData = new FormData()
  formData.append('file', file)
  const { data } = await api.post(`/contents/${contentId}/cover`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

// Collaborator management

export interface Collaborator {
  id: number
  username: string
  email: string
}

export async function listCollaborators(contentId: number): Promise<Collaborator[]> {
  const { data } = await api.get(`/contents/${contentId}/collaborators`)
  return data
}

export async function addCollaborator(contentId: number, userId: number): Promise<void> {
  await api.post(`/contents/${contentId}/collaborators/${userId}`)
}

export async function removeCollaborator(contentId: number, userId: number): Promise<void> {
  await api.delete(`/contents/${contentId}/collaborators/${userId}`)
}

export async function transferOwnership(contentId: number, newOwnerId: number): Promise<Content> {
  const { data } = await api.put(`/contents/${contentId}/owner/${newOwnerId}`)
  return data
}

// Calendar API

export async function fetchCalendarEvents(params: {
  start: string
  end: string
  status?: string
}): Promise<ContentCalendarItem[]> {
  const { data } = await api.get('/contents/calendar/events', { params })
  return data
}

export async function updateContentSchedule(
  id: number,
  scheduledPublishAt: string | null
): Promise<Content> {
  const { data } = await api.patch(`/contents/${id}/schedule`, {
    scheduled_publish_at: scheduledPublishAt,
  })
  return data
}

File: backend/app/schemas/content.py
from datetime import datetime

from pydantic import BaseModel


class ContentCreate(BaseModel):
    title: str
    content_markdown: str = ""
    content_html: str = ""
    source_type: str = "contribution"
    author: str = ""
    tags: list[str] = []
    category: str = ""
    cover_image: str | None = None
    scheduled_publish_at: datetime | None = None
    work_status: str = "planning"
    assignee_ids: list[int] = []
    # 多社区关联：不填则默认使用 X-Community-Id header 对应的社区
    community_ids: list[int] = []


class ContentUpdate(BaseModel):
    title: str | None = None
    content_markdown: str | None = None
    content_html: str | None = None
    source_type: str | None = None
    author: str | None = None
    tags: list[str] | None = None
    category: str | None = None
    cover_image: str | None = None
    scheduled_publish_at: datetime | None = None
    work_status: str | None = None
    assignee_ids: list[int] | None = None
    # 多社区关联：提供则替换全部关联；不提供则不变
    community_ids: list[int] | None = None


class ContentStatusUpdate(BaseModel):
    status: str


class ContentOut(BaseModel):
    id: int
    title: str
    content_markdown: str
    content_html: str
    source_type: str
    source_file: str | None
    author: str
    tags: list[str]
    category: str
    cover_image: str | None
    status: str
    work_status: str
    community_id: int | None
    created_by_user_id: int | None
    owner_id: int | None
    scheduled_publish_at: datetime | None
    created_at: datetime
    updated_at: datetime
    assignee_ids: list[int] = []
    # 多社区关联列表（由 API 层手动填充）
    community_ids: list[int] = []

    model_config = {"from_attributes": True}


class ContentListOut(BaseModel):
    id: int
    title: str
    source_type: str
    author: str
    tags: list[str]
    category: str
    status: str
    owner_id: int | None = None
    scheduled_publish_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ContentCalendarOut(BaseModel):
    """日历视图专用的轻量响应模型"""
    id: int
    title: str
    status: str
    source_type: str
    author: str
    category: str
    scheduled_publish_at: datetime | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ContentScheduleUpdate(BaseModel):
    """拖拽更新发布时间"""
    scheduled_publish_at: datetime | None = None


class PaginatedContents(BaseModel):
    items: list[ContentListOut]
    total: int
    page: int
    page_size: int
